sentry-android-core/src/main/java/io/sentry/android/core/DefaultAndroidEventProcessor.java

```diff
@@ -1,20 +1,769 @@
+    if (packageInfo != null) {
+      if (event.getRelease() == null) {
+        event.setRelease(packageInfo.packageName + "-" + packageInfo.versionName);
+      }
+      if (event.getDist() == null) {
+        event.setDist(getVersionCode(packageInfo));
+      }
+      if (event.getContexts().getApp() == null) {
+        event.getContexts().setApp(getApp(packageInfo));
+      }
+    }
+    setAppExtras(event);
+
+    // TODO: proguard UUIDs and debug-meta
+
+    if (event.getContexts().getDevice() == null) {
+      event.getContexts().setDevice(getDevice());
+    }
+    if (event.getContexts().getOperatingSystem() == null) {
+      event.getContexts().setOperatingSystem(getOperatingSystem());
+    }
+
+    // whats about Runtime, Browser, GPU object, do they make sense to Android?
+
+    //    GLESx always return null for those values
+    //    if (event.getContexts().getGpu() == null) {
+    //      Gpu gpu = new Gpu();
+    //      gpu.setApiType(GLES32.glGetString(GLES32.GL_RENDERER));
+    //      gpu.setVendorName(GLES32.glGetString(GLES32.GL_VENDOR));
+    //      gpu.setVersion(GLES32.glGetString(GLES32.GL_VERSION));
+    //      event.getContexts().setGpu(gpu);
+    //    }
+
     return event;
   }
+
+  private void setAppExtras(SentryEvent event) {
+    App app = event.getContexts().getApp();
+    if (event.getContexts().getApp() == null) {
+      app = new App();
+    }
+    app.setAppName(getApplicationName());
+    app.setAppStartTime(appStartTime);
+  }
+
+  private void setSdkVersion(SentryEvent event) {
+    SdkVersion sdkVersion = event.getSdkVersion();
+    if (sdkVersion == null) {
+      sdkVersion = new SdkVersion();
+    }
+    sdkVersion.setName("sentry-android");
+    String version = BuildConfig.VERSION_NAME; // do we need the version of each package?
+    sdkVersion.setVersion(version);
+    sdkVersion.addPackage("sentry-core", version);
+    sdkVersion.addPackage("sentry-android-core", version);
+    // sentry-android-ndk, integrations...
+  }
+
+  private String getVersionCode(PackageInfo packageInfo) {
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
+      return Long.toString(packageInfo.getLongVersionCode());
+    }
+    return getVersionCodeDep(packageInfo);
+  }
+
+  @SuppressWarnings("deprecation")
+  private String getVersionCodeDep(PackageInfo packageInfo) {
+    return Integer.toString(packageInfo.versionCode);
+  }
+
+  @SuppressWarnings("deprecation")
+  private String getAbi() {
+    return Build.CPU_ABI;
+  }
+
+  @SuppressWarnings("deprecation")
+  private String getAbi2() {
+    return Build.CPU_ABI2;
+  }
+
+  /**
+   * Return the Application's PackageInfo if possible, or null.
+   *
+   * @return the Application's PackageInfo if possible, or null
+   */
+  private PackageInfo getPackageInfo() {
+    try {
+      return context.getPackageManager().getPackageInfo(context.getPackageName(), 0);
+    } catch (Exception e) {
+      log(SentryLevel.ERROR, "Error getting package info.", e);
```

Comment:
That counts for the majority of occurrences of `SentryLevel.ERROR`, I would just log a warning instead of an error. Worst case a value in the context is missing but ERROR means really something critical.

Reply:
I'd say it's good as error. 
For _critical_ things we can use:

https://github.com/getsentry/sentry-android/blob/1bf3fed0ed92394ae8dd05db769e95674c6d3ac5/sentry-core/src/main/java/io/sentry/core/SentryLevel.java#L8
By default, we can have the SDK `debug=true` log out anything level `DEBUG` or higher. But for a dogfooding scenario or users interested in following internal errors of the SDK, they case raise the level to `ERROR` and get these logged out. This is how the .NET SDK works and it's helpful because users do report internal errors of the SDK which otherwise would go unnoticed.

We're stilll missing taking a SentryLevel for the internal logging in the Options object.

Reply:
I'd agree with @bruno-garcia if the other unified SDK already works like this.
@bruno-garcia should the log level be a task on asana?

Reply:
done